static/js/chart/draw_scatter.ts

```diff
@@ -550,6 +558,151 @@ function addRegressionLine(
+  yScale: ScatterScale,
+  quadrants: ChartQuadrant[]
+): Record<ChartQuadrant, Point[]> {
+  const numSectionsPerAxis = HIGHLIGHT_QUADRANT_SECTIONS * 2;
+  const xSectionLength =
+    (xScale.range()[1] - xScale.range()[0]) / numSectionsPerAxis;
+  const ySectionLength =
+    (yScale.range()[0] - yScale.range()[1]) / numSectionsPerAxis;
+  // Get the list of points in each section of the chart
+  const sectionPoints = {};
+  for (const point of points) {
+    const x = Math.floor(xScale(point.xVal) / xSectionLength);
+    const y = Math.floor(yScale(point.yVal) / ySectionLength);
+    if (!sectionPoints[x]) {
+      sectionPoints[x] = {};
+    }
+    if (!sectionPoints[x][y]) {
+      sectionPoints[x][y] = [];
+    }
+    sectionPoints[x][y].push(point);
+  }
+  // get the top (closest to the chart corner) MIN_HIGHLIGHT_POINTS points in
+  // each quadrant.
+  const result = {} as Record<ChartQuadrant, Point[]>;
+  for (const quadrant of quadrants) {
+    // depending on the quadrant traverse x and y axis either in increasing or
+    // decreasing value
+    const xIncrease =
+      quadrant == ChartQuadrant.TOP_RIGHT ||
+      quadrant == ChartQuadrant.BOTTOM_RIGHT;
+    const yIncrease =
+      quadrant == ChartQuadrant.BOTTOM_LEFT ||
+      quadrant == ChartQuadrant.BOTTOM_RIGHT;
+    // go layer by layer starting at the outside corner of each quadrant and add
+    // points until the whole quadrant has been searched or min number of points
+    // has been added.
+    const points: Point[] = [];
+    let layer = 0;
+    while (
+      layer < HIGHLIGHT_QUADRANT_SECTIONS &&
+      points.length < MIN_HIGHLIGHT_POINTS
+    ) {
+      const startingX = xIncrease ? numSectionsPerAxis - layer - 1 : layer;
+      const startingY = yIncrease ? numSectionsPerAxis - layer - 1 : layer;
+      // traverse along x of this section
+      if (xIncrease) {
+        for (let x = startingX; x < numSectionsPerAxis; x++) {
+          if (sectionPoints[x] && sectionPoints[x][startingY]) {
+            points.push(...sectionPoints[x][startingY]);
+          }
+        }
+      } else {
+        for (let x = startingX; x >= 0; x--) {
+          if (sectionPoints[x] && sectionPoints[x][startingY]) {
+            points.push(...sectionPoints[x][startingY]);
+          }
+        }
+      }
+      // traverse along y of this section
+      if (yIncrease) {
+        for (let y = startingY + 1; y < numSectionsPerAxis; y++) {
+          if (sectionPoints[startingX] && sectionPoints[startingX][y]) {
+            points.push(...sectionPoints[startingX][y]);
+          }
+        }
+      } else {
+        for (let y = startingY - 1; y >= 0; y--) {
+          if (sectionPoints[startingX] && sectionPoints[startingX][y]) {
+            points.push(...sectionPoints[startingX][y]);
+          }
+        }
+      }
+      layer++;
+    }
+    // filter out points that are too close to other points
+    result[quadrant] = points.filter((point, idx) => {
```

Comment:
hm, i wonder if we should reverse `points` before calling filter. If filter goes 0-n, then we might overfilter points that we'd want to keep (further towards the corners of the chart).

Reply:
actually, the points are in order of furthest towards the corners of the chart -> closer to the center of the chart!